Rewrote Prometheus Istio Setup Docs (#12157)
kyma-project/kyma: docs/01-overview/main-areas/observability/obsv-03-istio-monitoring.md

---
title: Monitoring in Istio
---

## Default setup

The [monitoring chart](https://github.com/kyma-project/kyma/blob/main/resources/monitoring/values.yaml) is preconfigured to collect all metrics relevant for observing the in-cluster [Istio](https://istio.io/latest/docs/concepts/observability/) Service Mesh, including the proxy-level, service-level, and control-plane metrics. The whole implementation of our monitoring solution is built around [Istio's observability best practices](https://istio.io/latest/docs/ops/best-practices/observability/).

![Prometheus Setup](./assets/prometheus-setup.svg)

1. The concept of collecting the [service-level](https://istio.io/latest/docs/concepts/observability/#service-level-metrics) metrics is based on the Istio Proxy implemented by Envoy. Istio Proxy collects all communication details inside the service mesh in a decentralized way. After scraping these high-cardinality metrics from the envoys, the metrics must be aggregated on a service level to get the final service-related details.

> 2. A dedicated Prometheus instance scrapes and aggregates the service-level metrics. That instance is configured with the smallest possible data retention time because the raw metrics scraped from the Istio Proxies have high-cardinality and don't need to be kept further. Instead, the main Prometheus instance scrapes the aggregated metrics through the `/federate` endpoint. This instance is a Deployment named `monitoring-prometheus-istio-server`, with a hardcoded configuration you should not change. It also has no PersistentVolume attached. This instance never discovers additional metric endpoints from such resources as ServiceMonitors.

Reviewer: "This instance is a Deployment". Which one? It is a bit confusing as you talked about the main instance in the sentence before.

> 2. A dedicated Prometheus instance scrapes and aggregates the service-level metrics. That instance is configured with the smallest possible data retention time because the raw metrics scraped from the Istio Proxies have high-cardinality and don't need to be kept further. Instead, the main Prometheus instance scrapes the aggregated metrics through the `/federate` endpoint. This instance is a Deployment named `monitoring-prometheus-istio-server`, with a hardcoded configuration you should not change. It also has no PersistentVolume attached. This instance never discovers additional metric endpoints from such resources as ServiceMonitors.

Reviewer 2: ```suggestion
2. A dedicated Prometheus instance (Prometheus-Istio) scrapes and aggregates the service-level metrics. That instance is configured with the smallest possible data retention time because the raw metrics scraped from the Istio Proxies have high-cardinality and don't need to be kept further. 
The Istio-Prometheus instance is a Deployment named `monitoring-prometheus-istio-server`, with a hardcoded configuration that must not be changed. It also has no PersistentVolume attached. This instance never discovers additional metric endpoints from such resources as ServiceMonitors.
```

> 3. The main Prometheus instance is configured to support scraping metrics using [`Strict mTLS`](https://istio.io/latest/docs/tasks/security/authentication/authn-policy/#globally-enabling-istio-mutual-tls-in-strict-mode). For this to work, Prometheus is configured to scrape metrics using Istio certificates. 

Reviewer: Shouldn't we mention here again that this instance now scrapes the istio-prometheus and everything else defined via a ServiceMonitor. Hereby, also mTLS based workload is supported

> 3. The main Prometheus instance is configured to support scraping metrics using [`Strict mTLS`](https://istio.io/latest/docs/tasks/security/authentication/authn-policy/#globally-enabling-istio-mutual-tls-in-strict-mode). For this to work, Prometheus is configured to scrape metrics using Istio certificates. 

Reviewer 2: ```suggestion
3. The main Prometheus instance scrapes these aggregated Istio metrics through the `/federate` endpoint of the Prometheus-Istio instance and any other metric endpoints from such resources as ServiceMonitors.
The main Prometheus instance supports scraping metrics using [`Strict mTLS`](https://istio.io/latest/docs/tasks/security/authentication/authn-policy/#globally-enabling-istio-mutual-tls-in-strict-mode). For this to work, Prometheus is configured to scrape metrics using Istio certificates.  
```

   
4. Prometheus is deployed with a sidecar proxy which rotates SDS certificates and outputs them to a volume mounted to the corresponding Prometheus container. It is configured to not intercept or redirect any traffic. 
   

> 5. By default, metrics from Kyma components are scraped using mTLS with some exceptions: Components whose workload is part of the Istio infrastructure, thus deployed with a sidecar proxy, are scraped using mTLS. Components deployed without sidecar proxy, are scraped using `http`, like controllers (e.g. Prometheus operator).

Reviewer 2: ```suggestion
5. By default, metrics from Kyma components are scraped using mTLS. The components' workload is part of the Istio infrastructure, they are deployed with a sidecar proxy, and scraped using mTLS.  As an exception, components deployed without sidecar proxy are scraped using `http` (for example, controllers like Prometheus operator).
```

I believe it's just that one exception, not "some"?

Author: Done

> 5. By default, metrics from Kyma components are scraped using mTLS. The components' workload is part of the Istio infrastructure, they are deployed with a sidecar proxy, and scraped using mTLS.  As an exception, components deployed without sidecar proxy are scraped using http (for example, controllers like Prometheus operator).

Reviewer: "The components' workload is part of the Istio infrastructure" - The workload is not part of the istio infrastructure in my pov.

> 5. By default, metrics from Kyma components are scraped using mTLS. The components' workload is part of the Istio infrastructure, they are deployed with a sidecar proxy, and scraped using mTLS.  As an exception, components deployed without sidecar proxy are scraped using http (for example, controllers like Prometheus operator).

Reviewer 2: ```suggestion
5. By default, metrics from Kyma components are scraped using mTLS. As an exception, components deployed without sidecar proxy are scraped using http (for example, controllers like Prometheus operator).
```

The Istio-related instance is a Deployment named `monitoring-prometheus-istio-server`, configured with a short data retention time and hardcoded configuration that you should not change. It also has no PersistentVolume attached. This instance never discovers additional metric endpoints from such resources as ServiceMonitors.

> The monitoring chart is configured in such a way that it is possible to scrape metrics using [`Strict mTLS`](https://istio.io/latest/docs/tasks/security/authentication/authn-policy/#globally-enabling-istio-mutual-tls-in-strict-mode). For this to work, Prometheus is configured to scrape metrics using Istio certificates. Prometheus is deployed with a sidecar proxy which rotates SDS certificates and outputs them to a volume mounted to the 

Reviewer 2: ```suggestion
The monitoring chart is configured to support scraping metrics using [`Strict mTLS`](https://istio.io/latest/docs/tasks/security/authentication/authn-policy/#globally-enabling-istio-mutual-tls-in-strict-mode). For this to work, Prometheus is configured to scrape metrics using Istio certificates. Prometheus is deployed with a sidecar proxy which rotates SDS certificates and outputs them to a volume mounted to the corresponding Prometheus container.
```

The monitoring chart is configured in such a way that it is possible to scrape metrics using [`Strict mTLS`]

> (https://istio.io/latest/docs/tasks/security/authentication/authn-policy/#globally-enabling-istio-mutual-tls-in-strict-mode). For this to work, Prometheus is configured to scrape metrics using Istio certificates. Prometheus is deployed with a sidecar proxy which rotates SDS certificates and outputs them to a volume mounted to the 

Reviewer 2: ```suggestion
(https://istio.io/latest/docs/tasks/security/authentication/authn-policy/#globally-enabling-istio-mutual-tls-in-strict-mode). For this to work, Prometheus is configured to scrape metrics using Istio certificates. Prometheus is deployed with a sidecar proxy which rotates SDS certificates and outputs them to a volume mounted to the corresponding Prometheus container. 
```

> corresponding Prometheus container. To stick to Istio's observability best practices, Prometheus's Istio-proxy is configured to not intercept or redirect any traffic. By default, metrics from Kyma components are scraped using mTLS with some exceptions. Components which's workload is part of the Istio infrastructure, thus deployed with a sidecar proxy, are scraped using mTLS. The others, deployed without sidecar proxy, are scraped using `http`, like controllers (e.g. Prometheus operator).

Reviewer 2: ```suggestion
Following Istio's observability best practices, Prometheus's Istio-proxy is configured to not intercept or redirect any traffic. By default, metrics from Kyma components are scraped using mTLS with some exceptions: Components whose workload is part of the Istio infrastructure, thus deployed with a sidecar proxy, are scraped using mTLS. Components deployed without sidecar proxy, are scraped using `http`, like controllers (e.g. Prometheus operator).
```
